# HG changeset patch
# User Andreas Weber <andy.weber.aw@gmail.com>
# Date 1471413723 -7200
#      Wed Aug 17 08:02:03 2016 +0200
# Node ID fc127f0e1ccde60ad5ad164df78b372c725216ec
# Parent  6d54ecbfb50a7be679264c57373c11dae4839a9d
Fix race condition when printing with Qt toolkit (#48519 and #44463)

ObjectProxy.cc: Use connect (..., Qt::BlockingQueuedConnection)
graphics.cc: Remove sleep

diff --git a/libgui/graphics/ObjectProxy.cc b/libgui/graphics/ObjectProxy.cc
--- a/libgui/graphics/ObjectProxy.cc
+++ b/libgui/graphics/ObjectProxy.cc
@@ -63,17 +63,18 @@ ObjectProxy::init (Object* obj)
         {
           connect (this, SIGNAL (sendUpdate (int)),
                    m_object, SLOT (slotUpdate (int)));
           connect (this, SIGNAL (sendFinalize (void)),
                    m_object, SLOT (slotFinalize (void)));
           connect (this, SIGNAL (sendRedraw (void)),
                    m_object, SLOT (slotRedraw (void)));
           connect (this, SIGNAL (sendPrint (const QString&, const QString&)),
-                   m_object, SLOT (slotPrint (const QString&, const QString&)));
+                   m_object, SLOT (slotPrint (const QString&, const QString&)),
+                   Qt::BlockingQueuedConnection);
         }
     }
 }
 
 void
 ObjectProxy::setObject (Object* obj)
 {
   emit sendFinalize ();
diff --git a/libinterp/corefcn/graphics.cc b/libinterp/corefcn/graphics.cc
--- a/libinterp/corefcn/graphics.cc
+++ b/libinterp/corefcn/graphics.cc
@@ -11101,17 +11101,23 @@ undocumented.
           // enough time to lock the mutex before we lock it here
           // again.  We thus wait 50 ms (which may not be enough) to
           // give it a chance: see http://octave.1599824.n4.nabble.com/Printing-issues-with-Qt-toolkit-tp4673270.html
 
           gh_manager::unlock ();
 
           go.get_toolkit ().print_figure (go, term, file, debug_file);
 
-          octave_sleep (0.05); // FIXME: really needed?
+          // FIXME: In ObjectProxy.cc ObjectProxy::init
+          // we now use connect (..., Qt::BlockingQueuedConnection)
+          // which should make the sleep unnecessary.
+          // See bug #44463 and #48519
+          // Remove it and the FIXME block above after testing.
+
+          // octave_sleep (0.05);
 
           gh_manager::lock ();
         }
     }
 
   return ovl ();
 }
 
